README.rst
====================
sshdefaultscan 0.4.0
====================

**Scan networks for SSH servers with default username and password.**

Use **sshdefaultscan** to scan networks or hosts for SSH servers, try to connect
using some default username and password. It uses `Nmap`_ to provide easy and
powerfull target selection and `Paramiko`_ to test credentials.

**Full documentation at** http://sshdefaultscan.readthedocs.org/

.. image:: https://readthedocs.org/projects/sshdefaultscan/badge/?version=latest


Usage
-----

Scan your own machine:

.. code-block:: bash

    $ python sshdefaultscan.py 127.0.0.1

    2015-06-08 21:16:57,711 - sshdefaultscan - DEBUG - Scanning...
    2015-06-08 21:17:03,892 - sshdefaultscan - DEBUG - 1 hosts up, 1 total in 0.28s
    2015-06-08 21:17:03,892 - sshdefaultscan - DEBUG - 127.0.0.1 Seems to have SSH open
    2015-06-08 21:17:06,001 - sshdefaultscan - INFO - 127.0.0.1 Logged in with root:root in 2.11s

Your local network, with ``--fast`` to improve speed:

.. code-block:: bash

    $ python sshdefaultscan.py --fast 192.168.1.1-254

    2015-06-08 21:21:59,408 - sshdefaultscan - DEBUG - Scanning...
    2015-06-08 21:22:08,807 - sshdefaultscan - DEBUG - 1 hosts up, 254 total in 3.38s
    2015-06-08 21:22:08,808 - sshdefaultscan - DEBUG - 192.168.1.42 Seems to have SSH open
    2015-06-08 21:22:11,463 - sshdefaultscan - DEBUG - 192.168.1.42 Authentication failed. (2.65s)

Different username or password:

.. code-block:: bash

    $ python sshdefaultscan.py --username admin --password 1234 192.168.1.1-254

    2015-06-08 21:21:59,408 - sshdefaultscan - DEBUG - Scanning...
    2015-06-08 21:22:08,807 - sshdefaultscan - DEBUG - 3 hosts up, 254 total in 3.11s
    2015-06-08 21:22:08,808 - sshdefaultscan - DEBUG - 192.168.1.42 Seems to have SSH open
    2015-06-08 21:22:11,463 - sshdefaultscan - INFO - 192.168.1.42  Logged in with admin:1234 in 0.98s

Or a much bigger network segment:

.. code-block:: bash

    $ python sshdefaultscan.py 192.168.99-110.1-254

    2015-06-08 21:24:47,177 - sshdefaultscan - DEBUG - Scanning...
    2015-06-08 21:25:16,035 - sshdefaultscan - DEBUG - 127 hosts up, 3048 total in 28.75s
    2015-06-08 21:25:16,035 - sshdefaultscan - DEBUG - 192.168.109.60 Seems to have SSH open
    2015-06-08 21:25:16,035 - sshdefaultscan - DEBUG - 192.168.110.182 Seems to have SSH open
    2015-06-08 21:25:16,035 - sshdefaultscan - DEBUG - 192.168.110.184 Seems to have SSH open
    2015-06-08 21:25:19,047 - sshdefaultscan - DEBUG - 192.168.109.60 Authentication failed. (3.01s)
    2015-06-08 21:25:20,436 - sshdefaultscan - DEBUG - 192.168.110.182 Authentication failed. (1.39s)
    ...

See full documentation at http://sshdefaultscan.readthedocs.org/


Disclaimer
----------

This software is provided for educational purposes and testing only: use it in
your own network or with permission from the network owner. I'm not responsible
of what actions people decide to take using this software. I'm not responsible
if someone do something against the law using this software. Please be good and
don't do anything harmful :)


Author
------

Andres Tarantini (atarantini@gmail.com)


License
-------

Released under GNU GPLv3, see `COPYING <https://github.com/atarantini/sshdefaultscan/blob/master/COPYING>`_ file for more details.

.. _Nmap: http://nmap.org/
.. _Paramiko: http://www.paramiko.org/
.. _`The Python Package Index`: https://pypi.python.org/pypi


docs/changelog.rst
Changelog
=========

0.4.0
-----

Documentation (2015-10-24)

    * Documentation moved to `docs` directory, implemented using `Sphinx`_.
    * Small refactoring and fixes over documentation.

0.3.0
-----

Custom Port (2015-09-17)

    * Added ``--port`` parameter to set custom SSH port.
    * Handle socket error when making SSH connection.

0.2.1
-----

Batch teamples (2015-07-03)

    * Batch mode custom output with ``--batch-template``.
    * Improved scan speed (in both normal and ``--fast``) by disabling reverse DNS resolution.

0.2.0
-----

Batch mode (2015-06-30)

    * Batch mode can be used with the ``--batch`` option.

0.1.3
-----

2015-06-19

  * Fixed logger: was using default username and password, now is using the ones sent by the user.

0.1.2
-----

Fast (2015-06-13)

  * Added ``--fast`` parameter to allow faster scans reducing timeouts (`T5 Nmap template <http://nmap.org/book/man-performance.html>`_).

0.1.1
-----

Custom credentials (2015-06-08)

  * Added ``--username`` and ``--password`` parameters to set default username and password.

0.1.0
-----

Initial (2015-06-07)

    * Initial release.

.. _Sphinx: http://sphinx-doc.org/

sshdefaultscan.py
#!/usr/bin/env python
"""
sshdefaultscan

Scan networks for SSH servers with default username and password.
"""
import argparse
import logging
import socket
from time import time

import nmap
import paramiko

SSH_DEFAULT_USERNAME = 'root'
SSH_DEFAULT_PASSWORD = 'root'
BATCH_TEMPLATE_DEFAULT = '{host}'


def out(hostname, username, password, port, template='{host}'):
    """
    Return a string to be used as output when "--batch" mode is enabled

    :param hostname:    String with the hostname
    :param username:    String with the username
    :param password:    String with the password
    :param template:    String with a template, defaults to "{host}" [1]

    :return:    String to be used as output

    [1] See https://docs.python.org/2/library/string.html#formatstrings
    """
    return template.format(
        host=hostname,
        username=username,
        password=password,
        port=port
    )

#
# Main
#
if __name__ == '__main__':
    ###########################################################################
    # Bootstrap
    #

    # Parse command line arguments
    parser = argparse.ArgumentParser(description='Scan networks for SSH servers with default username and password.')
    parser.add_argument('hosts', help='An IP address for a hostname or network, ex: 192.168.1.1 for single host or 192.168.1.1-254 for network.')
    parser.add_argument('--username', help='Set username, default is "root".', default=SSH_DEFAULT_USERNAME)
    parser.add_argument('--password', help='Set password, default is "root".', default=SSH_DEFAULT_PASSWORD)
    parser.add_argument('--port', help='Set port, default is 22.', default='22')
    parser.add_argument('--fast', help='Change timeout settings for the scanner in order to scan faster (T5).', default=False, action='store_true')
    parser.add_argument('--batch', help='Batch mode will only output hosts, handy to use with unix pipes.', default=False, action='store_true')
    parser.add_argument('--batch-template', help='Change batch mode output template, default is "{host}". Available context variables: host, username, password. Ex: "{username}@{host}" will return "root@192.168.0.1" as output when running in batch mode.', default=BATCH_TEMPLATE_DEFAULT)
    args = parser.parse_args()

    # If "--batch-template" is sent, assume that the user wants batch mode
    if args.batch_template != BATCH_TEMPLATE_DEFAULT:
        args.batch = True

    # Setup logging
    logger = logging.getLogger('sshdefaultscan')
    logger.setLevel(logging.DEBUG)
    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    file_handler = logging.FileHandler('sshdefaultscan.log')
    file_handler.setLevel(logging.INFO)
    file_handler.setFormatter(formatter)
    logger.addHandler(file_handler)
    if args.batch:
        logger.setLevel(logging.INFO)
    else:
        console_handler = logging.StreamHandler()
        console_handler.setLevel(logging.DEBUG)
        console_handler.setFormatter(formatter)
        logger.addHandler(console_handler)

    ###########################################################################
    # Scan
    #
    logger.debug('Scanning...')
    hosts = list()

    nmap_arguments = ['-n']
    if args.fast:
        nmap_arguments.append('-T5')
    nm = nmap.PortScanner()
    scan = nm.scan(args.hosts, args.port, arguments=' '.join(nmap_arguments))
    stats = scan.get('nmap').get('scanstats')
    logger.debug(
        '{up} hosts up, {total} total in {elapsed_time}s'.format(
            up=stats.get('uphosts'),
            total=stats.get('totalhosts'),
            elapsed_time=stats.get('elapsed')
        )
    )
    for host, data in list(scan.get('scan').items()):
        if data.get('tcp') and data.get('tcp').get(int(args.port)).get('state') == 'open':
            hosts.append(host)
            logger.debug('{host} Seems to have SSH open'.format(host=host))

    if not hosts:
        logger.debug('No hosts found with port {port} open.'.format(port=args.port))
        exit()

    ###########################################################################
    # Test credentials
    #
    logger.debug('Testing credentials...')
    for host in hosts:
        start_time = time()
        ssh = paramiko.SSHClient()
        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        try:
            ssh.connect(
                host,
                username=args.username,
                password=args.password,
                port=int(args.port)
            )

            if args.batch:
                print(out(host, args.username, args.password, args.port, template=args.batch_template))

            logger.info('{host} Logged in with {username}:{password} in {elapsed_time}s'.format(
                host=host,
                username=args.username,
                password=args.password,
                elapsed_time=round(time() - start_time, 2)
            ))
        except (
            paramiko.ssh_exception.AuthenticationException,
            paramiko.ssh_exception.SSHException,
            socket.error
        ) as e:
            logger.debug('{host} {exception} ({elapsed_time}s)'.format(
                host=host,
                exception=e,
                elapsed_time=round(time() - start_time, 2)
            ))
